File: KiCadDbLib/Projektanker.KiCad/KiCadLibraryReader.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Projektanker.KiCad
{
    public class KiCadLibraryReader
    {
        private string _getSymbol_library;
        private string[] _getSymbol_lines;

        public async Task<string> GetSymbolAsync(string library, string symbol, bool bufferLibrary = true)
        {
            if (!File.Exists(library))
            {
                throw new FileNotFoundException($"Symbol library \"{library}\" not found!");
            }
            string[] lines;

            if (bufferLibrary && (library == _getSymbol_library))
            {
                lines = _getSymbol_lines;
            }
            else
            {
                lines = await File.ReadAllLinesAsync(library);
                _getSymbol_library = bufferLibrary ? library : null;
                _getSymbol_lines = bufferLibrary ? lines : null;
            }

            int start = 0, end = 0;
            // Search begin of symbol (DEF ...)
            for (int i = 0; i < lines.Length; i++)
            {
                var line = lines[i];
                if (!line.StartsWith($"DEF {symbol}")) continue;

                start = i;
                break;
            }

            if (start == 0)
            {
                throw new KeyNotFoundException($"Symbol \"{symbol}\" not found in symbol library \"{library}\"");
            }
            // Search end of symbol (ENDDEF)
            for (int i = start + 1; i < lines.Length; i++)
            {
                var line = lines[i];
                if (!line.StartsWith("ENDDEF")) continue;

                end = i + 1;
                break;
            }

            if (end < start)
            {
                throw new Exception($"Symbol library \"{library}\" is corrupted! \"ENDDEF\" not found!");
            }

            // Return symbol
            StringBuilder stringBuilder = new StringBuilder();
            // Add comment before symbol
            stringBuilder.AppendLine($"# {symbol}");
            stringBuilder.AppendLine("#");

            for (int i = start; i < end; i++)
            {
                stringBuilder.AppendLine(lines[i]);
            }
            // Add comment after symbol
            stringBuilder.AppendLine("#");

            return stringBuilder.ToString();
        }

        public async Task<string> GetSymbolTemplateAsync(KiCadPart part)
        {
            var split = part.Symbol.Split(":").ToList();
            if (split.Count < 2)
            {
                throw new FormatException("Symbol must be defined as \"path:symbol\"");
            }

            string symbol = split[split.Count-1];
            split.RemoveAt(split.Count - 1);
            string path = string.Join(':', split);

            // Get symbol
            var template = await GetSymbolAsync(path, symbol);

            var lines = template.Split(Environment.NewLine, StringSplitOptions.RemoveEmptyEntries).ToList();
            // Remove lines starting with # ( and add them later )
            while (lines[0].StartsWith("#"))
            {
                lines.RemoveAt(0);
            }

            // Prepare for string.Format()
            int i = 0;
            int format, startIndex, index;

            string start;
            char find;
            // "DEF value reference ..." to "DEF {0} {1} ..."
            start = "DEF "; format = 0; find = ' ';
            startIndex = start.Length;
            index = lines[i].IndexOf(find, startIndex);
            lines[i] = lines[i].Remove(startIndex, index - startIndex);
            lines[i] = lines[i].Insert(startIndex, "{" + format + "}");
            start = "DEF {0} "; format = 1;
            startIndex = start.Length;
            index = lines[i].IndexOf(find, startIndex);
            lines[i] = lines[i].Remove(startIndex, index - startIndex);
            lines[i] = lines[i].Insert(startIndex, "{" + format + "}");

            // Clear fields F0 - F3
            start = "Fx \"";
            find = '"';
            for (i = 1; i < 5; i++)
            {
                switch (i)
                {
                    case 1: format = 1; break;
                    case 2: format = 0; break;
                    case 3: format = 2; break;
                    case 4: format = 3; break;
                    default:
                        break;
                }
                startIndex = start.Length;
                index = lines[i].IndexOf(find, startIndex);
                lines[i] = lines[i].Remove(startIndex, index - startIndex);
                lines[i] = lines[i].Insert(startIndex, "{" + format + "}");
            }

            // Remove F4 and following
            while (lines[i].StartsWith("F"))
            {
                lines.RemoveAt(i);
            }

            // Insert {4} for custom fields
            lines.Insert(i, "{4}");

            // Find and remove ALIAS
            for (i = 0; i < lines.Count; i++)
            {
                if (!lines[i].StartsWith("ALIAS")) continue;

                lines.RemoveAt(i);
                break;
            }

            // Restore deleted #-Lines
            lines.Insert(0, "# {0}");
            lines.Insert(1, "#");

            return string.Join(Environment.NewLine, lines);
        }

        public async Task<IList<string>> GetSymbolsAsync(string library)
        {
            if (!File.Exists(library))
            {
                throw new FileNotFoundException($"File {library} not found!");
            }

            List<string> symbols = new List<string>();

            using (var file = File.OpenText(library))
            {
                while (file.Peek() > 0)
                {
                    var line = await file.ReadLineAsync();
                    if (!line.StartsWith("DEF")) continue;

                    int start, end;
                    start = "DEF ".Length;
                    end = line.IndexOf(' ', start);
                    symbols.Add(line.Substring(start, end - start));
                }
            }
            return symbols;
        }
    }
}

File: electron/src/kicad/kicad-library-reader.ts
import { fs } from '../fs';
import { Part } from '../parts/part';

export class KiCadLibraryReader {
  private _getSymbol_library: string;
  private _getSymbol_lines: string[];

  public async getSymbolAsync(
    library: string,
    symbol: string,
    bufferLibrary: boolean = true
  ): Promise<string> {
    try {
      if (!fs.checkFileExists(library)) {
        throw new Error(`Symbol library \"${library}\" not found!`);
      }

      var lines: string[];

      if (bufferLibrary && library == this._getSymbol_library) {
        lines = this._getSymbol_lines;
      } else {
        var text = await fs.readFile(library, 'utf8');
        lines = text.replace('\r\n', '\n').split('\n');
        this._getSymbol_library = bufferLibrary ? library : null;
        this._getSymbol_lines = bufferLibrary ? lines : null;
      }

      var start = 0,
        end = 0;
      // Search begin of symbol (DEF ...)
      for (var i = 0; i < lines.length; i++) {
        const line = lines[i];
        if (!line.startsWith(`DEF ${symbol}`)) continue;

        start = i;
        break;
      }

      if (start == 0) {
        throw new Error(
          `Symbol \"${symbol}\" not found in symbol library \"${library}\"`
        );
      }
      // Search end of symbol (ENDDEF)
      for (var i = start + 1; i < lines.length; i++) {
        var line = lines[i];
        if (!line.startsWith('ENDDEF')) continue;

        end = i + 1;
        break;
      }

      if (end < start) {
        throw new Error(
          `Symbol library \"${library}\" is corrupted! \"ENDDEF\" not found!`
        );
      }

      // Return symbol
      var stringBuilder: string[] = [];
      // Add comment before symbol
      stringBuilder.push(`"# ${symbol}`);
      stringBuilder.push('#');

      for (i = start; i < end; i++) {
        stringBuilder.push(lines[i]);
      }
      // Add comment after symbol
      stringBuilder.push('#');

      return Promise.resolve(stringBuilder.join('\n'));
    } catch (error) {
      return Promise.reject(error);
    }
  }

  public async getSymbolTemplateAsync(part: Part): Promise<string> {
    try {
      var split = part.symbol.split(':');
      if (split.length < 2) {
        throw new Error('Symbol must be defined as "path:symbol"');
      }

      var symbol = split[split.length - 1];
      split.pop();
      var path = split.join(':');

      // Get symbol
      var template = await this.getSymbolAsync(path, symbol);

      // Normalize new line, split on new line and remove empty lines
      var lines = template
        .replace('\r\n', '\n')
        .split('\n')
        .filter(l => {
          return l;
        });
      // Remove lines starting with # ( and add them later )
      while (lines[0].startsWith('#')) {
        lines.shift();
      }

      // Prepare for string.Format()
      let i = 0;
      var format, startIndex, index;

      var start: string;
      var find;
      // "DEF value reference ..." to "DEF {0} {1} ..."
      start = 'DEF ';
      format = 0;
      find = ' ';
      startIndex = start.length;
      index = lines[i].indexOf(find, startIndex);
      lines[i] =
        lines[i].substr(0, startIndex) + `{${format}}` + lines[i].substr(index);
      start = 'DEF {0} ';
      format = 1;
      startIndex = start.length;
      index = lines[i].indexOf(find, startIndex);
      lines[i] =
        lines[i].substr(0, startIndex) + `{${format}}` + lines[i].substr(index);

      // Clear fields F0 - F3
      start = 'Fx "';
      find = '"';
      for (i = 1; i < 5; i++) {
        switch (i) {
          case 1:
            format = 1;
            break;
          case 2:
            format = 0;
            break;
          case 3:
            format = 2;
            break;
          case 4:
            format = 3;
            break;
          default:
            break;
        }
        startIndex = start.length;
        index = lines[i].indexOf(find, startIndex);
        lines[i] =
          lines[i].substr(0, startIndex) +
          `{${format}}` +
          lines[i].substr(index);
      }

      // Remove F4 and following
      while (lines[i].startsWith('F')) {
        lines.splice(i, 1);
      }

      // Insert {4} for custom fields
      lines.splice(i, 0, '{4}');

      // Find and remove ALIAS
      for (i = 0; i < lines.length; i++) {
        if (!lines[i].startsWith('ALIAS')) continue;

        lines.splice(i, 1);
        break;
      }

      // Restore deleted #-Lines
      lines.splice(0, 0, '# {0}');
      lines.splice(1, 0, '#');

      return Promise.resolve(lines.join('\n'));
    } catch (error) {
      return Promise.reject(error);
    }
  }

  public async getSymbolsAsync(library: string): Promise<string[]> {
    try {
      if (!fs.checkFileExists(library)) {
        throw new Error(`File \"${library}\" not found!`);
      }

      var symbols = [];
      var text = await fs.readFile(library, 'utf8');
      var lines = text.replace('\r\n', '\n').split('\n');
      for (let i = 0; i < lines.length; i++) {
        const line = lines[i];
        if (!line.startsWith('DEF')) continue;
        var start, end;
        start = 'DEF '.length;
        end = line.indexOf(' ', start);
        symbols.push(line.substring(start, end));
      }

      return Promise.resolve(symbols);
    } catch (error) {
      Promise.reject(error);
    }
  }
}
